File: packages/js/benchmarks/pricing.ts
import assert from 'node:assert/strict'
import { performance } from 'node:perf_hooks'
import process from 'node:process'
import { parseArgs } from 'node:util'

import type { ModelPrice, ModelPriceCalculationResult, PriceCalculationResult, Provider, Usage } from '../src/types'

import { calcPrice as calcPricePublic, updatePrices } from '../src/api'
import { calcPrice as calcPriceDirect } from '../src/engine'

const DEFAULT_ITERATIONS = 10_000
const DEFAULT_SAMPLES = 5
const DEFAULT_WARMUP_ITERATIONS = 2_000
const BENCHMARK_TIMESTAMP = new Date('2026-01-01T00:00:00Z')

interface BenchmarkCase {
  expected: ModelPriceCalculationResult
  modelPrice: ModelPrice
  modelRef: string
  name: string
  usage: Usage
}

interface BenchmarkResult {
  caseName: string
  maxNsPerOp: number
  medianNsPerOp: number
  minNsPerOp: number
  pathName: string
}

interface BenchmarkSettings {
  iterations: number
  samples: number
  warmupIterations: number
}

function priceResult(result: PriceCalculationResult): ModelPriceCalculationResult {
  assert.ok(result, 'public price calculation did not find the benchmark fixture')
  return {
    input_price: result.input_price,
    output_price: result.output_price,
    total_price: result.total_price,
  }
}

function assertPriceResult(actual: ModelPriceCalculationResult, expected: ModelPriceCalculationResult, label: string): void {
  for (const field of ['input_price', 'output_price', 'total_price'] as const) {
    const difference = Math.abs(actual[field] - expected[field])
    assert.ok(difference <= Number.EPSILON, `${label} ${field} ${actual[field].toString()} != ${expected[field].toString()}`)
  }
}

function createCases(): BenchmarkCase[] {
  const complexUsage: Usage = {
    cache_audio_read_tokens: 100,
    cache_read_tokens: 400,
    input_audio_tokens: 300,
    input_tokens: 1_000,
    output_audio_tokens: 50,
    output_tokens: 200,
  }
  const complexExpected = {
    input_price: 0.00816,
    output_price: 0.0056,
    total_price: 0.01376,
  }
  const complexBuiltIn = calcPricePublic(complexUsage, 'gpt-realtime', {
    providerId: 'openai',
    timestamp: BENCHMARK_TIMESTAMP,
  })
  assertPriceResult(priceResult(complexBuiltIn), complexExpected, 'complex built-in result changed')
  assert.ok(complexBuiltIn)

  const fixtures: Omit<BenchmarkCase, 'modelRef'>[] = [
    {
      expected: { input_price: 0, output_price: 0, total_price: 0 },
      modelPrice: {},
      name: 'empty',
      usage: {},
    },
    {
      expected: { input_price: 0.002, output_price: 0, total_price: 0.002 },
      modelPrice: { input_mtok: 2 },
      name: 'one-key',
      usage: { input_tokens: 1_000 },
    },
    {
      expected: { input_price: 0.002, output_price: 0.001, total_price: 0.003 },
      modelPrice: { input_mtok: 2, output_mtok: 10 },
      name: 'ordinary-two-key',
      usage: { input_tokens: 1_000, output_tokens: 100 },
    },
    {
      expected: { input_price: 0.001575, output_price: 0, total_price: 0.001575 },
      modelPrice: {
        cache_audio_read_mtok: 0.25,
        cache_read_mtok: 0.5,
        input_audio_mtok: 3,
        input_mtok: 2,
      },
      name: 'four-key-cache-modality-overlap',
      usage: {
        cache_audio_read_tokens: 100,
        cache_read_tokens: 400,
        input_audio_tokens: 300,
        input_tokens: 1_000,
      },
    },
    {
      expected: complexExpected,
      modelPrice: complexBuiltIn.model_price,
      name: 'complex-six-key-built-in',
      usage: complexUsage,
    },
  ]

  return fixtures.map((fixture) => ({ ...fixture, modelRef: `benchmark-${fixture.name}` }))
}

function createBenchmarkProvider(cases: BenchmarkCase[]): Provider {
  return {
    api_pattern: '^benchmark://',
    id: 'benchmark',
    models: cases.map((benchmarkCase) => ({
      id: benchmarkCase.modelRef,
      match: { equals: benchmarkCase.modelRef },
      prices: benchmarkCase.modelPrice,
    })),
    name: 'Local benchmark fixture',
  }
}

function assertExpectedResults(benchmarkCase: BenchmarkCase): void {
  const direct = calcPriceDirect(benchmarkCase.usage, benchmarkCase.modelPrice)
  assertPriceResult(direct, benchmarkCase.expected, `${benchmarkCase.name} direct result changed`)

  const publicCalculation = calcPricePublic(benchmarkCase.usage, benchmarkCase.modelRef, {
    providerId: 'benchmark',
    timestamp: BENCHMARK_TIMESTAMP,
  })
  assertPriceResult(priceResult(publicCalculation), benchmarkCase.expected, `${benchmarkCase.name} public result changed`)
}

function runIterations(operation: () => unknown, iterations: number): void {
  for (let iteration = 0; iteration < iterations; iteration += 1) {
    operation()
  }
}

function measure(
  caseName: string,
  pathName: string,
  operation: () => unknown,
  { iterations, samples, warmupIterations }: BenchmarkSettings
): BenchmarkResult {
  runIterations(operation, warmupIterations)

  const sampleNsPerOp: number[] = []
  for (let sample = 0; sample < samples; sample += 1) {
    const started = performance.now()
    runIterations(operation, iterations)
    const elapsedNs = (performance.now() - started) * 1_000_000
    sampleNsPerOp.push(elapsedNs / iterations)
  }
  sampleNsPerOp.sort((left, right) => left - right)

  const middle = Math.floor(sampleNsPerOp.length / 2)
  const minNsPerOp = sampleNsPerOp[0]
  const maxNsPerOp = sampleNsPerOp.at(-1)
  const upperMiddle = sampleNsPerOp[middle]
  assert.ok(minNsPerOp !== undefined && maxNsPerOp !== undefined && upperMiddle !== undefined)

  let medianNsPerOp = upperMiddle
  if (sampleNsPerOp.length % 2 === 0) {
    const lowerMiddle = sampleNsPerOp[middle - 1]
    assert.ok(lowerMiddle !== undefined)
    medianNsPerOp = (lowerMiddle + upperMiddle) / 2
  }

  return {
    caseName,
    maxNsPerOp,
    medianNsPerOp,
    minNsPerOp,
    pathName,
  }
}

function runBenchmarks(settings: BenchmarkSettings): BenchmarkResult[] {
  const cases = createCases()
  const provider = createBenchmarkProvider(cases)
  updatePrices(({ setProviderData }) => {
    setProviderData([provider])
  })

  for (const benchmarkCase of cases) {
    assertExpectedResults(benchmarkCase)
  }

  const results: BenchmarkResult[] = []
  for (const benchmarkCase of cases) {
    results.push(measure(benchmarkCase.name, 'direct', () => calcPriceDirect(benchmarkCase.usage, benchmarkCase.modelPrice), settings))
    results.push(
      measure(
        benchmarkCase.name,
        'public',
        () =>
          calcPricePublic(benchmarkCase.usage, benchmarkCase.modelRef, {
            providerId: 'benchmark',
            timestamp: BENCHMARK_TIMESTAMP,
          }),
        settings
      )
    )
  }
  return results
}

function positiveInt(value: string, option: string): number {
  const parsed = Number(value)
  if (!Number.isSafeInteger(parsed) || parsed < 1) {
    throw new Error(`${option} must be a positive integer`)
  }
  return parsed
}

function parseSettings(): BenchmarkSettings {
  const { values } = parseArgs({
    options: {
      iterations: { default: String(DEFAULT_ITERATIONS), type: 'string' },
      samples: { default: String(DEFAULT_SAMPLES), type: 'string' },
      'warmup-iterations': { default: String(DEFAULT_WARMUP_ITERATIONS), type: 'string' },
    },
  })

  return {
    iterations: positiveInt(values.iterations, '--iterations'),
    samples: positiveInt(values.samples, '--samples'),
    warmupIterations: positiveInt(values['warmup-iterations'], '--warmup-iterations'),
  }
}

function main(): void {
  const settings = parseSettings()
  const results = runBenchmarks(settings)

  console.log(`Node ${process.version} (V8 ${process.versions.v8})`)
  console.log(
    `iterations=${settings.iterations.toString()} samples=${settings.samples.toString()} warmup_iterations=${settings.warmupIterations.toString()}`
  )
  console.log('case                                      path      median ns/op      min ns/op      max ns/op')
  for (const result of results) {
    console.log(
      `${result.caseName.padEnd(41)} ${result.pathName.padEnd(7)} ${result.medianNsPerOp.toFixed(1).padStart(13)} ${result.minNsPerOp.toFixed(1).padStart(14)} ${result.maxNsPerOp.toFixed(1).padStart(14)}`
    )
  }
}

main()

File: benchmarks/python/pricing.py
from __future__ import annotations

import argparse
import platform
import statistics
import sys
from collections.abc import Callable, Sequence
from dataclasses import dataclass
from datetime import datetime, timezone
from decimal import Decimal
from time import perf_counter_ns

from genai_prices import Usage, calc_price
from genai_prices.data_snapshot import DataSnapshot, get_snapshot, set_custom_snapshot
from genai_prices.types import CalcPrice, ClauseEquals, ModelInfo, ModelPrice, PriceCalculation, Provider

DEFAULT_ITERATIONS = 10_000
DEFAULT_SAMPLES = 5
DEFAULT_WARMUP_ITERATIONS = 2_000
BENCHMARK_TIMESTAMP = datetime(2026, 1, 1, tzinfo=timezone.utc)

PriceResult = tuple[Decimal, Decimal, Decimal]


@dataclass(frozen=True)
class BenchmarkCase:
    name: str
    usage: Usage
    model_price: ModelPrice
    expected: PriceResult

    @property
    def model_ref(self) -> str:
        return f'benchmark-{self.name}'


@dataclass(frozen=True)
class BenchmarkResult:
    case_name: str
    path_name: str
    median_ns_per_op: float
    min_ns_per_op: float
    max_ns_per_op: float


def create_cases() -> tuple[BenchmarkCase, ...]:
    _, complex_model = get_snapshot().find_provider_model('gpt-realtime', None, 'openai', None)
    complex_price = complex_model.get_prices(BENCHMARK_TIMESTAMP)

    return (
        BenchmarkCase(
            name='empty',
            usage=Usage(),
            model_price=ModelPrice(),
            expected=(Decimal(0), Decimal(0), Decimal(0)),
        ),
        BenchmarkCase(
            name='one-key',
            usage=Usage(input_tokens=1_000),
            model_price=ModelPrice(input_mtok=Decimal(2)),
            expected=(Decimal('0.002'), Decimal(0), Decimal('0.002')),
        ),
        BenchmarkCase(
            name='ordinary-two-key',
            usage=Usage(input_tokens=1_000, output_tokens=100),
            model_price=ModelPrice(input_mtok=Decimal(2), output_mtok=Decimal(10)),
            expected=(Decimal('0.002'), Decimal('0.001'), Decimal('0.003')),
        ),
        BenchmarkCase(
            name='four-key-cache-modality-overlap',
            usage=Usage(
                input_tokens=1_000,
                cache_read_tokens=400,
                input_audio_tokens=300,
                cache_audio_read_tokens=100,
            ),
            model_price=ModelPrice(
                input_mtok=Decimal(2),
                cache_read_mtok=Decimal('0.5'),
                input_audio_mtok=Decimal(3),
                cache_audio_read_mtok=Decimal('0.25'),
            ),
            expected=(Decimal('0.001575'), Decimal(0), Decimal('0.001575')),
        ),
        BenchmarkCase(
            name='complex-six-key-built-in',
            usage=Usage(
                input_tokens=1_000,
                cache_read_tokens=400,
                input_audio_tokens=300,
                cache_audio_read_tokens=100,
                output_tokens=200,
                output_audio_tokens=50,
            ),
            model_price=complex_price,
            expected=(Decimal('0.00816'), Decimal('0.0056'), Decimal('0.01376')),
        ),
    )


def create_benchmark_snapshot(cases: Sequence[BenchmarkCase]) -> DataSnapshot:
    models = [
        ModelInfo(
            id=case.model_ref,
            match=ClauseEquals(case.model_ref),
            prices=case.model_price,
        )
        for case in cases
    ]
    provider = Provider(
        id='benchmark',
        name='Local benchmark fixture',
        api_pattern=r'^benchmark://',
        models=models,
    )
    return DataSnapshot(providers=[provider], from_auto_update=False)


def direct_result(result: CalcPrice) -> PriceResult:
    return result['input_price'], result['output_price'], result['total_price']


def public_result(result: PriceCalculation) -> PriceResult:
    return result.input_price, result.output_price, result.total_price


def assert_expected_results(case: BenchmarkCase) -> None:
    direct = direct_result(case.model_price.calc_price(case.usage))
    assert direct == case.expected, f'{case.name} direct result {direct!r} != {case.expected!r}'

    public = public_result(
        calc_price(
            case.usage,
            case.model_ref,
            provider_id='benchmark',
            genai_request_timestamp=BENCHMARK_TIMESTAMP,
        )
    )
    assert public == case.expected, f'{case.name} public result {public!r} != {case.expected!r}'


def run_iterations(operation: Callable[[], object], iterations: int) -> None:
    for _ in range(iterations):
        operation()


def measure(
    case_name: str,
    path_name: str,
    operation: Callable[[], object],
    *,
    iterations: int,
    samples: int,
    warmup_iterations: int,
) -> BenchmarkResult:
    run_iterations(operation, warmup_iterations)

    sample_ns_per_op: list[float] = []
    for _ in range(samples):
        started = perf_counter_ns()
        run_iterations(operation, iterations)
        elapsed = perf_counter_ns() - started
        sample_ns_per_op.append(elapsed / iterations)

    return BenchmarkResult(
        case_name=case_name,
        path_name=path_name,
        median_ns_per_op=statistics.median(sample_ns_per_op),
        min_ns_per_op=min(sample_ns_per_op),
        max_ns_per_op=max(sample_ns_per_op),
    )


def run_benchmarks(*, iterations: int, samples: int, warmup_iterations: int) -> list[BenchmarkResult]:
    cases = create_cases()
    set_custom_snapshot(create_benchmark_snapshot(cases))

    try:
        for case in cases:
            assert_expected_results(case)

        results: list[BenchmarkResult] = []
        for case in cases:
            results.append(
                measure(
                    case.name,
                    'direct',
                    lambda case=case: case.model_price.calc_price(case.usage),
                    iterations=iterations,
                    samples=samples,
                    warmup_iterations=warmup_iterations,
                )
            )
            results.append(
                measure(
                    case.name,
                    'public',
                    lambda case=case: calc_price(
                        case.usage,
                        case.model_ref,
                        provider_id='benchmark',
                        genai_request_timestamp=BENCHMARK_TIMESTAMP,
                    ),
                    iterations=iterations,
                    samples=samples,
                    warmup_iterations=warmup_iterations,
                )
            )
        return results
    finally:
        set_custom_snapshot(None)


def positive_int(value: str) -> int:
    parsed = int(value)
    if parsed < 1:
        raise argparse.ArgumentTypeError('must be at least 1')
    return parsed


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description='Benchmark Python pricing calculation overhead.')
    parser.add_argument('--iterations', type=positive_int, default=DEFAULT_ITERATIONS)
    parser.add_argument('--samples', type=positive_int, default=DEFAULT_SAMPLES)
    parser.add_argument('--warmup-iterations', type=positive_int, default=DEFAULT_WARMUP_ITERATIONS)
    return parser.parse_args()


def main() -> None:
    args = parse_args()
    results = run_benchmarks(
        iterations=args.iterations,
        samples=args.samples,
        warmup_iterations=args.warmup_iterations,
    )

    implementation = platform.python_implementation()
    print(f'Python {platform.python_version()} ({implementation})')
    print(f'iterations={args.iterations} samples={args.samples} warmup_iterations={args.warmup_iterations}')
    print('case                                      path      median ns/op      min ns/op      max ns/op')
    for result in results:
        print(
            f'{result.case_name:<41} {result.path_name:<7} '
            f'{result.median_ns_per_op:>13.1f} {result.min_ns_per_op:>14.1f} {result.max_ns_per_op:>14.1f}'
        )


if __name__ == '__main__':
    sys.exit(main())
